bin/monitor.ml

```diff
@@ -42,7 +42,17 @@ module Event = struct
   let progress p = Progress p
 end
 
-module State = struct
+module State : sig
+  (** Internal state of the render loop together with updates that update the UI via side
+      effects. *)
+
+  type t
+
+  val init : unit -> t
+  val update_jobs : t -> Job.Event.t list -> unit
+  val update_progress : t -> Progress.t -> unit
+  val update_diagnostics : t -> Diagnostic.Event.t list -> unit
```

Comment:
This function is still weird. Update and render should be separate steps and rendering should be done once after all the updates.

```
(match event with
| .. -> State.update state ..);
State.render state
```

Reply:
If this pattern isn't efficient enough, you can make the update functions return some `update` value to help do the update more efficiently and make render:

```
type update
val render : t -> update option -> unit
```

Reply:
I've made an incremental update type for the state that gets returned by these update functions.